File: src/tracker.rs
use std::fs::File;
use std::io::{self, Read, Write};
use serde_json;
use crate::contribution::Contribution;

pub struct Tracker {
    contributions: Vec<Contribution>,
}

impl Tracker {
    pub fn new() -> Self {
        let mut tracker = Tracker {
            contributions: Vec::new(),
        };
        tracker.load_contributions().ok(); // Try to load contributions
        tracker
    }

    // Add a contribution
    pub fn add_contribution(&mut self, user: &str, project: &str, contribution_type: &str) {
        let contribution = Contribution::new(user, project, contribution_type);
        self.contributions.push(contribution);
        self.save_contributions().unwrap();
    }

    // List all contributions
    pub fn list_contributions(&self) {
        if self.contributions.is_empty() {
            println!("No contributions yet.");
        } else {
            for contribution in &self.contributions {
                println!(
                    "User: {} | Project: {} | Contribution: {}",
                    contribution.user, contribution.project, contribution.contribution_type
                );
            }
        }
    }

    // Save contributions to a JSON file
    pub fn save_contributions(&self) -> io::Result<()> {
        let file = File::create("contributions.json")?;
        serde_json::to_writer(file, &self.contributions)?;
        Ok(())
    }

    // Load contributions from a JSON file
    pub fn load_contributions(&mut self) -> io::Result<()> {
        let mut file = File::open("contributions.json")?;
        let mut data = String::new();
        file.read_to_string(&mut data)?;
        self.contributions = serde_json::from_str(&data)?;
        Ok(())
    }
}

File: assignment_10/tracker.py
import json
import os
from datetime import datetime

DATA_FILE = "contributions.json"

# Load existing data or create new
def load_contributions():
    if os.path.exists(DATA_FILE):
        with open(DATA_FILE, "r") as file:
            return json.load(file)
    else:
        return []

# Save to JSON
def save_contributions(data):
    with open(DATA_FILE, "w") as file:
        json.dump(data, file, indent=4)

# Add new contribution
def add_contribution():
    repo = input("Repository name: ")
    date = input("Date (YYYY-MM-DD): ")
    message = input("Commit message: ")
    try:
        commits = int(input("Number of commits: "))
    except ValueError:
        print("Invalid number of commits.")
        return

    contribution = {
        "repository": repo,
        "date": date,
        "message": message,
        "commits": commits
    }

    contributions = load_contributions()
    contributions.append(contribution)
    save_contributions(contributions)
    print("✅ Contribution added!\n")

# View all contributions
def view_contributions():
    contributions = load_contributions()
    if not contributions:
        print("No contributions found.\n")
        return

    for i, c in enumerate(contributions, 1):
        print(f"{i}. {c['date']} - {c['repository']} ({c['commits']} commits)")
        print(f"   Message: {c['message']}")
    print()

# View summary
def view_summary():
    contributions = load_contributions()
    total_commits = sum(c["commits"] for c in contributions)
    unique_repos = set(c["repository"] for c in contributions)
    print(f"📦 Total contributions: {len(contributions)}")
    print(f"📈 Total commits: {total_commits}")
    print(f"📁 Repositories involved: {len(unique_repos)}\n")

# Menu
def main():
    while True:
        print("=== Open Source Tracker ===")
        print("1. Add contribution")
        print("2. View all contributions")
        print("3. View summary")
        print("4. Exit")
        choice = input("Choose an option: ")

        if choice == "1":
            add_contribution()
        elif choice == "2":
            view_contributions()
        elif choice == "3":
            view_summary()
        elif choice == "4":
            print("👋 Goodbye!")
            break
        else:
            print("❌ Invalid choice. Try again.\n")

if __name__ == "__main__":
    main()

